varlink: implement file descriptor passing

poettering opened the pull request:
This is split out/preparation for #26826. 

This has no users right now, but regardless of #26826 fd passing via varlink definitely makes sense either way. Hence I think this is safe to merge early.

Note that the official varlink specs don't define fd passing. We implement this as trivial extension to the protocol: if the program enables fd passing (separately for input and output) then we'll read any fds we get along with messages and pass them to the app.

There's no feature negotiation or anything. We do however validate that the used transport is AF_UNIX when enabling the concept.

keszybz reviewed (changes requested):
Looks reasonable. Two questions about design:
1. why enqueue the fds on Varlink object, instead of extending the call methods to accept an additional `int[]` to pass everything in one go?
2. as an alternative to having a queue of json objects, why not serialize everything and keep a sequence of offsets where there are message boundaries with fds to apply?

poettering commented:
> why enqueue the fds on Varlink object, instead of extending the call methods to accept an additional int[] to pass everything in one go?

I had it like that initially, actually. And then reworked it like this.

The main point why I reworked this goes something like this: in my usecase for this I need to do a bunch of different things (allocate user namespace fd, allocate multiple mount fds, allocate a tracking pipe fd), and they are relatively independent. Once I have done that all I send a combined message back with one json object that is made up of the results of all these separate operations plus an fd array that is made up of all these fds. And this is where it becomes tricky: we somehow need to reference the right fd from the json object. The obvious way is via indexes, i.e. encode an integer in json that delares "the n-th fd passed along with this message is the mount fd for /usr" and so on. But this doesn't work if we still have to combine the individual fd arrays from the multiple suboperations into one, because that would mean we'd have to retroactively change the json data the individual operations gave us with the right overall indexes. So instead I figured, we need something that tracks the index numbers for us – that I can pass into the first operation, and then into the second operation, and the third, and that each time returns the next available index into the message. And I just decided to make this the varlink conneciton object, since that's kinda natural then.

Or in other words: it's really nice if we can build up our messages piecemeal: accumulate a json object here, and fd there, and then combine them, and finally fire them off, without having to keep complex state in the client that allows us to keep them together. With the implemented approach things are easy: whenever i want to attach an fd to the json, i just pass it to the varlink object and attach the index it returns instead. Problem solved.

Does that make sense?

keszybz commented on src/test/test-varlink.c:
```diff
@@ -48,6 +49,55 @@ static int method_something(Varlink *link, JsonVariant *parameters, VarlinkMetho
 }
 
+static void test_fd(int fd, const void *buf, size_t n) {
+        _cleanup_free_ void *rbuf = NULL;
+        ssize_t m;
+
+        rbuf = malloc(n + 1);
+        assert_se(rbuf);
```
```C
int rbuf[n + 1];
```

This is test code, no need to make it overcomplicated.

keszybz commented on src/test/test-varlink.c:
```diff
@@ -151,11 +203,40 @@ static void *thread(void *arg) {
+        assert_se(fd2 >= 0);
+        assert_se(fd3 >= 0);
+
+        assert_se(varlink_push_fd(c, fd1) >= 0);
+        assert_se(varlink_push_fd(c, fd2) >= 0);
+        assert_se(varlink_push_fd(c, fd3) >= 0);
```
This should probably test the return number, i.e. `== 0`, `== 1`, `== 2`.

keszybz commented:
Hmm, that actually raises more questions. Because if you're building up the message piecemeal, then presumably the steps can fail. And with fds attached to the Varlink object, suddenly cleanup becomes unclear: what do you do it you decide *not* to send the message after all?

poettering commented on src/shared/varlink.c:
```diff
@@ -81,6 +81,15 @@ typedef enum VarlinkState {
+struct VarlinkJsonQueue {
+        LIST_FIELDS(VarlinkJsonQueue, queue);
+        JsonVariant *data;
+        size_t n_fds;
+        int fds[];
```
yes, will add comment

poettering commented:
> Hmm, that actually raises more questions. Because if you're building up the message piecemeal, then presumably the steps can fail. And with fds attached to the Varlink object, suddenly cleanup becomes unclear: what do you do it you decide not to send the message after all?

This is indeed a good point I didn't think about so far.

Here's what I added now to address this: a new call varlink_reset_fds() that allows flushing out all currently queued fds from the "push_fds" array. Clients can call that when they realize the message they were building ultimately didn't work out. Now, this of course is a bit fragile: people are going to forget to call this. Hence I added one more thing: whenever an error reply message is sent we'll implicitly call this new function, under the assumption that it never makes sense to send an fd along with an error code... This should at least cover all my own current codepaths nicely. 

The alternative would be to build up a new kind of fd array object in parallel to the json object, and then pass that along with the json submission, but this sounded a bit over the top to me, given the fd passing is likely not going to be the common case, but the more corner case.

poettering commented:
> 2\. as an alternative to having a queue of json objects, why not serialize everything and keep a sequence of offsets where there are message boundaries with fds to apply?

Yeah, so I was thinking about this for a while, but I feared that these offsets would be fragile. It's kinda what makes C suck so hard: dealing with buffer offsets incorrectly after all. So this approach of just storing a linked list queue of the messages to send plus their fds appeared nicer to me.

But this is nicer for other reasons too: json variants are to a large degree about sharing memory, i.e. they are refcounted individually and down the tree. Thus it made sense to me to stick to that as long as we can, and convert them to string form only at the last step, where we cannot share the memory anymore. (I am not sure this actually will save too much memory, since in most cases we'll probably have a refcount of 1 on things, but it still *felt* better ;-).

Also, json objects carry various metadata, such as debug info (i.e. parsing location), and a senseitive boolean for securely erasing them from memory after use. I think it's nice to keep that info around as long as we can to make debugging easier, and so that we can honour this metadata as long as we can.

(that said: the "sensitive" flag is currently not honoured by the varlink code once serialized into the text buffer. We probably should though, but maybe that's material for another patch...)

poettering commented:
Force pushed a new version. Addressed all raised points. I also added one commit on top, to follow up on this:

> (that said: the "sensitive" flag is currently not honoured by the varlink code once serialized into the text buffer. We probably should though, but maybe that's material for another patch...)

i.e. the new commit makes minor changes to propagate the sensitive flag into the serialized output buffer, and then erases it securely after it passed the buffer to the kernel.

That commit only covers the sensitive flag for the output buffer. I figured we should honour it given we have it. We should add a concept for setting it on input too, but I left that for a later PR, when we actually need it. In that case I think we should just allow marking the whole varlink connection as "sensitive" in which case we'll set the flag for all incoming json objects automatically, wholesale.

keszybz commented:
OK, makes sense. Thanks for the detailed explanations. 

(The solution with `varlink_reset_fds()` should work fine in practice. Another approach would be to extend the json library code to allow attaching additional metadata to the json fields. For example, a label and a cleanup function. Then the varlink code could look for some agreed-upon label that would mean that a integer is an fd, and treat it accordingly. It would unset the label after taking ownership of the fd. And if that never happened, the cleanup function would close the fd. I think this could be fun and versatile. But OTOH, it seems like quite a bit of work, so it doesn't seem worth to implement it just for this one use case. Maybe if more cases pop up.)
